compiler/rustc_mir/src/transform/function_references.rs

```diff
@@ -0,0 +1,158 @@
+            fn_span: _,
+        } = &terminator.kind
+        {
+            let span = match func {
+                Operand::Copy(place) | Operand::Move(place) => {
+                    self.body.local_decls[place.local].source_info.span
+                }
+                Operand::Constant(constant) => constant.span,
+            };
+            self.calls.push(span);
+        };
+        self.super_terminator(terminator, location);
+    }
+    fn visit_rvalue(&mut self, rvalue: &Rvalue<'tcx>, location: Location) {
+        match rvalue {
+            Rvalue::Ref(_, _, place) | Rvalue::AddressOf(_, place) => {
+                let decl = &self.body.local_decls[place.local];
+                if let ty::FnDef(def_id, _) = decl.ty.kind {
+                    let ident = self
+                        .body
+                        .var_debug_info
+                        .iter()
+                        .find(|info| info.source_info.span == decl.source_info.span)
+                        .map(|info| info.name.to_ident_string())
+                        .unwrap_or(self.tcx.def_path_str(def_id));
+                    let lint = FunctionRefLint { ident, def_id, source_info: self.source_info };
+                    self.potential_lints.push(lint);
+                }
+            }
+            Rvalue::Cast(_, op, _) => {
+                let op_ty = op.ty(self.body, self.tcx);
+                if self.is_fn_ref(op_ty) {
+                    self.casts.push(self.source_info.span);
+                }
+            }
+            _ => {}
+        }
+        self.super_rvalue(rvalue, location);
+    }
+}
+
+impl<'a, 'tcx> FunctionRefChecker<'a, 'tcx> {
+    fn is_fn_ref(&self, ty: Ty<'tcx>) -> bool {
+        let referent_ty = match ty.kind {
+            ty::Ref(_, referent_ty, _) => Some(referent_ty),
+            ty::RawPtr(ty_and_mut) => Some(ty_and_mut.ty),
+            _ => None,
+        };
+        referent_ty
+            .map(|ref_ty| if let ty::FnDef(..) = ref_ty.kind { true } else { false })
+            .unwrap_or(false)
+    }
+}
+
+struct FunctionRefLint {
+    ident: String,
+    def_id: DefId,
+    source_info: SourceInfo,
+}
+
+impl<'tcx> FunctionRefLint {
+    fn emit(&self, tcx: TyCtxt<'tcx>, body: &Body<'tcx>) {
+        let def_id = self.def_id;
+        let source_info = self.source_info;
+        let lint_root = body.source_scopes[source_info.scope]
+            .local_data
+            .as_ref()
+            .assert_crate_local()
+            .lint_root;
+        let fn_sig = tcx.fn_sig(def_id);
+        let unsafety = fn_sig.unsafety().prefix_str();
+        let abi = match fn_sig.abi() {
+            Abi::Rust => String::from(""),
+            other_abi => {
+                let mut s = String::from("extern \"");
+                s.push_str(other_abi.name());
+                s.push_str("\" ");
+                s
+            }
+        };
+        let num_args = fn_sig.inputs().map_bound(|inputs| inputs.len()).skip_binder();
```

Comment:
Another minor question, is `skip_binder()` appropriate here and on line 144 in this context?